File: attribouter/src/main/java/me/jfenn/attribouter/adapters/WedgeAdapter.java
package me.jfenn.attribouter.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import me.jfenn.attribouter.wedges.Wedge;

public class WedgeAdapter extends RecyclerView.Adapter<Wedge.ViewHolder> {

    private List<Wedge> wedges;

    public WedgeAdapter(List<Wedge> wedges) {
        this.wedges = wedges;
    }

    @NonNull
    @Override
    public Wedge.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Wedge info = wedges.get(viewType);
        return info.getViewHolder(LayoutInflater.from(parent.getContext()).inflate(info.getLayoutRes(), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull Wedge.ViewHolder holder, int position) {
        wedges.get(position).bind(holder.itemView.getContext(), holder);
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return wedges.size();
    }

}

File: attribouter/src/main/java/me/jfenn/attribouter/adapters/WedgeAdapter.kt
package me.jfenn.attribouter.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.jfenn.attribouter.wedges.Wedge

class WedgeAdapter(private val wedges: List<Wedge<*>>) : RecyclerView.Adapter<Wedge.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Wedge.ViewHolder {
        val info = wedges[viewType]
        return info.getViewHolder(LayoutInflater.from(parent.context).inflate(info.layoutRes, parent, false))
    }

    override fun onBindViewHolder(holder: Wedge.ViewHolder, position: Int) {
        (wedges[position] as? Wedge<Wedge.ViewHolder>)?.bind(holder.itemView.context, holder)
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    override fun getItemCount(): Int {
        return wedges.size
    }

}